networkx/utils/decorators.py

```diff
@@ -392,84 +355,868 @@ def random_array(dims, random_state=1):
     --------
     py_random_state
     """
-
-    @decorator
-    def _random_state(func, *args, **kwargs):
-        # Parse the decorator arguments.
-        try:
-            random_state_arg = args[random_state_index]
-        except TypeError as e:
-            raise nx.NetworkXError("random_state_index must be an integer") from e
-        except IndexError as e:
-            raise nx.NetworkXError("random_state_index is incorrect") from e
-
-        # Create a numpy.random.RandomState instance
-        random_state = create_random_state(random_state_arg)
-
-        # args is a tuple, so we must convert to list before modifying it.
-        new_args = list(args)
-        new_args[random_state_index] = random_state
-        return func(*new_args, **kwargs)
-
-    return _random_state
+    return argmap(create_random_state, random_state_argument)
 
 
 np_random_state = random_state
 
 
-def py_random_state(random_state_index):
+def py_random_state(random_state_argument):
     """Decorator to generate a random.Random instance (or equiv).
 
-    Argument position `random_state_index` processed by create_py_random_state.
+    The decorator processes the argument indicated by `random_state_argument`
+    using :func:`nx.utils.create_py_random_state`.
+    The argument value can be a seed (integer), or a random number generator.
```

Comment:
I added more to maybe make this clearer, but there were 9 cases to cover so it might be too-much-info.

Reply:
Yikes, yes I see!  Now I wonder: how many of those cases do we use and need?

For the current implementation, the docs are good.  But the implementation itself may be too magical.  Maybe that's for another PR.

Reply:
Yes -- that is probably a good PR for v3.0.  We included all those versions because at the time we had code that used each of those mechanisms. Maybe that has changed...  Our goal was to make it as easy as possible for people to submit code to the project. At the same time we wanted to let users provide a single RNG for everything. So we constructed an interface to make the `random.Random` API available even with a numpy RandomState RNG passed in.